### ISSUE TITLE
How to train node classification using custom dataset

### ISSUE THREAD

**Issue by vigneshgig** [stellargraph/stellargraph#1661]
HI, thanks for the great library,
I have seen the loading pandas jupyter to generate the node classification dataset, but I confused with the train example of node classification (gat and graphsage). For my understanding CORA network dataset has only single graph input. So how can I train multiple input graph Let's say I have 10 invoice or id card image, So using OCR I able to get the bounding box of text Now I want to classify the bounding box whether it is a company name, address, etc. based on the location of structure so that's why I planned to use GCN technique.
This is my dataset look like:
invoice_1:
source   |   target   |   distance feature | isalpha feature | label
0                 1                 xxx                       1                      header
0                 2                  xxx                       0                     bill no
0                 3                 xxx                        1                     company_name
3                 7                xxx                          0                    date
invoice_2:
source   |   target   |   distance feature | isalpha feature | label
0                 1                 xxx                       1                      header
0                 2                  xxx                       0                     bill no
0                 4                 xxx                        1                     company_name
4                 5                 xxx                        1                     company_name

and so on
first two invoice_2 is the same billing but ocr bounding box will differ if invoice image varies like first invoice photo taken near so the ocr detect the company service as a single bounding box but in second invoice photo taken far so, the ocr detect the company separate and service as a separate bounding box. 
So How can I train the gat or graphsage with this dataset which as a separate CSV file for each invoice image?

I inspired by this https://github.com/dhavalpotdar/Graph-Convolution-on-Structured-Documents GitHub link to write a code for invoice and id card to my use case and your loading pandas homogenous graph example to generate above file CSV.
I hope that the above explanation will be understandable.
Thanks @huonw

**Comment by vigneshgig** [comment]
Can any help me on this

**Comment by kjun9** [comment]
Hi @vigneshgig thanks for checking out StellarGraph, that sounds like an interesting use-case for graph neural networks! I'll try my best to answer your question, but I'm not sure if I understand the dataset you've described, so I might be able to give better suggestions if you can clarify some things I'm confused about.

So each invoice appears to be a set of edges (containing `source` and `target` columns) joining two entities in a document, however the example values you've given in the `label` column (eg. "header", "date") sound more like it belongs to an entity rather than a relationship between two entities? Could you clarify what each row and each column in the invoice dataset represents?

> So How can I train the gat or graphsage with this dataset which as a separate CSV file for each invoice image?

In general, in order to use methods like GAT/GraphSAGE/GCN which operate on a single graph, you must construct a single instance of the `StellarGraph` class with the information across all your CSV files combined.

If the dataset only contains edges *within* each invoice, it'll just mean that your full graph consists of communities which are not connected to each other. As an example:

```python
import pandas as pd

# read individual CSV invoice files
edges_from_invoice_1 = pd.read_csv(...)
edges_from_invoice_2 = pd.read_csv(...)

# create one big dataframe with all edges
edges_all = pd.concat([edges_from_invoice_1, edges_from_invoice_2])

graph = StellarGraph(edges=edges_all)
``` 

However doing this naively means **you'll be treating each node ID from all invoices as the same node**. So with the example data that you've provided, the resulting set of edges will be:
```
src, tgt
0, 1
0, 2
0, 3
3, 7
0, 1
0, 2
0, 4
4, 5
```
which means "node 0" has 6 outgoing edges in total. If this is not what you want, and instead you want each invoice to be completely independent of each other, then you might want to prepend the node IDs that occur in each invoice with some kind of unique identifier for the invoice. For example:
```
src, tgt
inv1_0, inv1_1
inv1_0, inv1_2
inv1_0, inv1_3
inv1_3, inv1_7
inv2_0, inv2_1
inv2_0, inv2_2
inv2_0, inv2_4
inv2_4, inv2_5
```

Based on the [blog post](https://towardsdatascience.com/using-graph-convolutional-neural-networks-on-structured-documents-for-information-extraction-c1088dcd2b8f) by the author of the github repository you've linked, it seems like node features containing the location of each node is also important for this task. This means you also need a separate dataframe containing a row for each node in the graph.:
```
id, x_min, x_max, y_min, y_max
inv1_0, 12, 34, 1, 5
inv1_1, ...
...
inv2_0, ...
inv2_1, ...
...
```
and construct the StellarGraph using both dataframes:
```python
import pandas as pd

# read individual CSV files that contain edge information for each invoice
edges_from_invoice_1 = pd.read_csv(...)
edges_from_invoice_2 = pd.read_csv(...)

# read individual CSV files that contain node feature information for each invoice
nodes_from_invoice_1 = pd.read_csv(...)
nodes_from_invoice_2 = pd.read_csv(...)

# create two big dataframes - one for all nodes, another for all edges
edges_all = pd.concat([edges_from_invoice_1, edges_from_invoice_2])
nodes_all = pd.concat([nodes_from_invoice_1, nodes_from_invoice_2])

graph = StellarGraph(nodes=nodes_all, edges=edges_all)
```

Currently the only type of methods that support a series of graphs as input are for graph classification - you can take a look at [this demo](https://stellargraph.readthedocs.io/en/stable/demos/graph-classification/gcn-supervised-graph-classification.html) for example. It seems like you're looking to do node classification so it may not be all that relevant, but it might be worth a look if you think there's a possibility that you could re-frame your problem as graph classification problem.

Hope that helps!

**Comment by vigneshgig** [comment]
Thanks for the reply, I already tried, what you said above mention process using graphsaga,  I have tried different parameter, different node feature and also edge feature.  but I unable to get output properly, the loss is getting stuck in-between the range. As you said maybe this does not support this kind of sceneries, Is graphSage  support edge weight? if not so, please suggest any other graph network which supports edge weight. 
Please look at the below jupyter pdf which I had been tested.
[graph_2.pdf](https://github.com/stellargraph/stellargraph/files/4750361/graph_2.pdf)
](url)
Note: I used id card as a dataset
Dataset Explanation:
I have taken the bounding box of text and I have calculated center of the bounding box then, I used an algorithm to get up left right down nearby four bounding box for a pivoted bounding box then, I select another bounding box as a pivot and do the same process to connect the box like that node connection graph is generated.
Feature :  isalpha which contains 0 or 1 regarding the text contains any number or not
For example: 
Vignesh - 0  name
09/34/323 - 1 dob
AT3484 - 1  id
unwanted other text marked as a header, sub_header, or footer as per the location of the text.
and distance between the connected bounding box as edge weight or feature

First Try:
Node feature: isalpha
Edge: I used distance columns a weight for an edge, but I didn't get output well.
Second Try:
I have increased the layer size but the same result
Third Try:
Node feature: isalpha, distance
Edge: None Because I looked that graphsage does not support edge weight.
same result
Fourth Try:
Node feature: isalpha, distance
Edge weight:None
Edge Feature: Distance
same result
Fifty Try:
It is same above the case but I normalized every column. but same result
Sixth Try:
Node Feature: isalpha, distance, centroid_x, centroid_y
Edge Feature: None
or 
Edge Feature: distance
same result,  In t-SNE visualization, the embedding is get spread all over the scatter graph, but for above 1,2,3,4 process, the embedding is not spread all over the scatter graph(but followed some pattern).

Please look at the pdf So you can get some idea what I explained above process.
I am looking for any suggestion,
Thanks,

**Comment by vigneshgig** [comment]
If possible please tell me a hint for converting graph classification into the node classification. @kjun9 @huonw

**Comment by vigneshgig** [comment]
I find this article for GCN to classify the text box. https://nanonets.com/blog/information-extraction-graph-convolutional-networks/. @kjun9 @huonw 
Thanks

**Comment by vigneshgig** [comment]
Node classification is generating an embedding vector 16D for each node using aggregate and then we send it to the denser layer to classify the each node. So we can't send a separate separate each graph for node classification and then for graph classification it kind similar to convolutional neural network image classification. If we want to do like graph classification as a node classification we have to do like GCN Auto encoder or pixel-wise prediction in image segmentation like output have to be same as input. 
And the above article nanonets I think they used only single invoice graph to train the node classification, not like multiple invoice graphs for node classification which I was used to training node classification by combining the multiple graphs.
If I am wrong please correct for node classification and graph classification

**Comment by kjun9** [comment]
Hi @vigneshgig are you able to share your actual `.ipynb` file? The PDF is a little difficult to read due to the connection error overlay and some cells are cut off:
<img width="1436" alt="image" src="https://user-images.githubusercontent.com/33508488/84217301-244b5f80-ab0f-11ea-998c-1926f8356a40.png">
Or you can convert the `.ipynb` file to a PDF using: 
```
jupyter nbconvert --to pdf <file>
```

From what I could see, this plot looks like you might be overtraining:
<img width="1354" alt="image" src="https://user-images.githubusercontent.com/33508488/84217472-9754d600-ab0f-11ea-9bb1-efc71b027c20.png">
Around the 26th epoch, the loss for the validation set starts to increase, so you probably want to stop training around there. The embeddings look okay and colours do seem to be forming clusters, but is likely to improve further if you implement the early stopping to avoid overtraining.

Otherwise I think the rest of the approach looks pretty reasonable!

> Is graphSage support edge weight? if not so, please suggest any other graph network which supports edge weight.

GraphSAGE doesn't support edge weights - you can take a look at the [table we have in our docs](https://stellargraph.readthedocs.io/en/stable/demos/index.html#find-a-demo-for-an-algorithm), there is a column for "edge weights" so any of the algorithms in that table with a "yes" in that column do support edge weights.

I would probably suggest trying GCN - our demo notebook for GCN doesn't use edge weights, but the option is available when constructing a `FullBatchGenerator(..., weighted=True)` ([link to docs](https://stellargraph.readthedocs.io/en/stable/api.html#stellargraph.mapper.FullBatchNodeGenerator))

**Comment by huonw** [comment]
> GraphSAGE doesn't support edge weights - you can take a look at the table we have in our docs, there is a column for "edge weights" so any of the algorithms in that table with a "yes" in that column do support edge weights.

(Relevant issue #462)

**Comment by vigneshgig** [comment]
Can you tell me how to implement gcn autoencoder using graph classification model @huonw @kjun9  because I felt Current node classification implementation is not right choose for this use case because even I stop at 26th epochs  Accuracy is not good. I send the a pdf because ipyn suddenly got an issue. That's why i have taken screen shot. So I try send new one.

Try to check other network which has edge weight support then I will let you know
Thanks

**Comment by huonw** [comment]
Have you seen the demos of unsupervised representation learning https://stellargraph.readthedocs.io/en/stable/demos/embeddings/index.html ? It's not quite an autoencoder, but may still solve the task you're interested in.

>  felt Current node classification implementation is not right choose for this use case because even I stop at 26th epochs Accuracy is not good

Could you be more specific about what "accuracy is not good" means? Do you have a baseline model you're comparing against? Or maybe a published paper that suggests what is possible? Without a reference point, I think it's hard to say whether an accuracy of 0.55 (estimated from the screenshot) is good or not.

> I send the a pdf because ipyn suddenly got an issue. That's why i have taken screen shot. So I try send new one.

For sharing Jupyter notebooks, good options include:

- using the built-in PDF export (or even the markdown export) as Kevin points out
- copying the file contents into a [Gist](https://gist.github.com) and sharing a link to that, because GitHub will render it nicely right in the web browser. (There's various tools that make this easy from the command line, too, like https://github.com/mbostock/gistup or https://github.com/defunkt/gist.)

> @huonw @kjun9 

Just a little thing: we're happy to help, but we do get notifications for every comment even without any `@`s, and will respond promptly without them too.

**Comment by vigneshgig** [comment]
Sorry for the late update, I have tried the unsupervised and also weight_cross_entropty loss function with GCN and graphsage, but the result is not good to my use case. My intuition is, only multi graph node classification should work for this sceneries. So I working on combining graph classification code and node classification code. For a multigraph, we need padding to make every graph structure equal size and using the mask to ignore the padded node which was implemented in the graph classification and In node classification, we get an individual latent represented vector for each node and using it we predicting the class for each node, so In graph classification from the pooling layer, I want to get the output as a  N no of  node individual vector from that I want to connect each vector to the dense layer and predict the class like pixelwise classification.
So How can I add padding technique in node classification gcn.
![download](https://user-images.githubusercontent.com/34392627/84652108-b8129680-af28-11ea-8007-85582e59808b.png)

any suggestion,
Thanks

**Comment by vigneshgig** [comment]
Hi, I created the graph and node classification by modifying the code. Now I seeing the results.
Thanks,

**Comment by Sharathmk99** [comment]
@vigneshgig can you explain more on how did you combine graph and node classification? Can you share the results?

**Comment by vigneshgig** [comment]
@Sharathmk99  Hi, Sry still testing is going on, So soon I will try to give the code because after finishing the company product, I can give the code and also testing somewhat satisfy compare to single graph node classification, the only problem I am facing is overfitting. I got a 98% percentage of accuracy but the testing result is 50% to 60%, I just trained with only 80 id card graphs, So I planning to try it on a larger dataset. Once I got the result, I will send the code.

**Comment by Sharathmk99** [comment]
@vigneshgig Sure. Thank you

**Comment by vigneshgig** [comment]
Hi, I finally excited to say that I got the result very well, and also I just tested with variate of the dataset. I just did some modification on the stellar graph then I used some new feature engineering and also augmentation technique to give better results. The cool thing about the project was with a single image we can augment the multi -dataset to get the best result. So soon I will release the stellar modification code, feature engineering code, and also augmentation code, If you ok with it.
@kjun9 
Thanks for your kick start.

**Comment by ysfali** [comment]
@vigneshgig is it possible for you to share some kind of flow of how your final algorithm looked like? I'm working on a similar problem related to Billing documents and would like some hints to push me in the right direction. Thanks in advance.
